### PR TITLE
feat: electron support

### LINKED ISSUE #3412: support electron-renderer target
### What problem does this feature solve?

support build electron application using rspack

### What does the proposed API of configuration look like?

just like webpack https://webpack.js.org/configuration/target/#:~:text=electron%5B%5BX%5D.Y%5D%2Drenderer
it seems we only need to port this plugin to rspack https://github.com/webpack/webpack/blob/main/lib/electron/ElectronTargetPlugin.js

### PR DESCRIPTION
**Pull request by ahabhgk**
## Related issue (if exists)

closes #3412

### <samp>🤖 Generated by Copilot at 191a058</samp>

The pull request improves the target and externals options for the rspack tool and the rspack core and plugin crates. It simplifies the target logic by removing unused code and refactoring the target parsing and validation. It implements JSONP chunk loading for dynamic imports in the test fixture. It adds support for externalizing common electron modules in different contexts based on the target and externals presets. It updates the documentation, schema, and tests for the target and externals options.

<details open=true>
  <summary><h2>Walkthrough</h2></summary>

### <samp>🤖 Generated by Copilot at 191a058</samp>

*  Add electron target plugin and presets to externalize common electron modules ([link](https://github.com/web-infra-dev/rspack/pull/3414/files?diff=unified&w=0#diff-c8d061ed21897caebe63ce5a159b65df3dfd38620f4fd4a902504c59bf40f052R1-R71), [link](https://github.com/web-infra-dev/rspack/pull/3414/files?diff=unified&w=0#diff-675991f6e765627c5acbd6484c811589d6fc5f6d2998e65c535cb27f9f340936L3-R8), [link](https://github.com/web-infra-dev/rspack/pull/3414/files?diff=unified&w=0#diff-6112b39f2c1def8a0f89824413cfd6936f187aa6ca3712e1e6aeb4d34f67ae41R154-R181), [link](https://github.com/web-infra-dev/rspack/pull/3414/files?diff=unified&w=0#diff-5f5bb1aeb2a977ff30c5ab822c79f09b76476c503462939c69beb9e4d7552123L217-R221), [link](https://github.com/web-infra-dev/rspack/pull/3414/files?diff=unified&w=0#diff-55e0fbd74437dff4e251152f5d93efd7f5871cbad72a7959dddb6add8fbf60ccR545-R570), [link](https://github.com/web-infra-dev/rspack/pull/3414/files?diff=unified&w=0#diff-d7e2c50a4845569f35f387513dc5a6e542e9d5dde107294fd55c5ff4e8a72c88R462-R486), [link](https://github.com/web-infra-dev/rspack/pull/3414/files?diff=unified&w=0#diff-790d97d1c7c2bfbbe7720a38355ec1e0d462206a5ddb9826c6aa8bff7090ce0fR468-R471), [link](https://github.com/web-infra-dev/rspack/pull/3414/files?diff=unified&w=0#diff-8c9c55e119547fb481071dee846c579a8df388a933cd5628c0249875644a8478L6-R11))
*  Change target type and schema to include more node and electron variants ([link](https://github.com/web-infra-dev/rspack/pull/3414/files?diff=unified&w=0#diff-790d97d1c7c2bfbbe7720a38355ec1e0d462206a5ddb9826c6aa8bff7090ce0fL368-R381), [link](https://github.com/web-infra-dev/rspack/pull/3414/files?diff=unified&w=0#diff-e63a3b8814d8c67470d1e315c43b58f648e3e3b22a266598a23a5d2a44dbf2a7L3-R56))
*  Remove platform field and logic from target struct and test config module ([link](https://github.com/web-infra-dev/rspack/pull/3414/files?diff=unified&w=0#diff-4232ded1f7ad0508146d85a38c706a1493b225c379a187ddcfe67a2700430655L4-R5), [link](https://github.com/web-infra-dev/rspack/pull/3414/files?diff=unified&w=0#diff-4232ded1f7ad0508146d85a38c706a1493b225c379a187ddcfe67a2700430655L39), [link](https://github.com/web-infra-dev/rspack/pull/3414/files?diff=unified&w=0#diff-4232ded1f7ad0508146d85a38c706a1493b225c379a187ddcfe67a2700430655L45), [link](https://github.com/web-infra-dev/rspack/pull/3414/files?diff=unified&w=0#diff-4232ded1f7ad0508146d85a38c706a1493b225c379a187ddcfe67a2700430655L74-R58), [link](https://github.com/web-infra-dev/rspack/pull/3414/files?diff=unified&w=0#diff-8b7ae8eae42b4d9202dbd35407c3338295c42d2b3b863c1a9c179c23502c86c2L9-R9), [link](https://github.com/web-infra-dev/rspack/pull/3414/files?diff=unified&w=0#diff-8b7ae8eae42b4d9202dbd35407c3338295c42d2b3b863c1a9c179c23502c86c2L549-R552))
*  Add TODO comment to move target option out of compiler options ([link](https://github.com/web-infra-dev/rspack/pull/3414/files?diff=unified&w=0#diff-b6c971a9adb8c246d789a436e09e6bb0f0077c819d437857845e60263aedf103R13))
*  Add new fields to raw external struct for electron externals presets ([link](https://github.com/web-infra-dev/rspack/pull/3414/files?diff=unified&w=0#diff-b3d4e4741437a4af781bf0c7ffc91117b42dcefe17fd5ad19fb175e303124332R176-R179))
*  Remove extra asterisk from mode section comment ([link](https://github.com/web-infra-dev/rspack/pull/3414/files?diff=unified&w=0#diff-790d97d1c7c2bfbbe7720a38355ec1e0d462206a5ddb9826c6aa8bff7090ce0fL68-R68))
*  Add test case for electron renderer target ([link](https://github.com/web-infra-dev/rspack/pull/3414/files?diff=unified&w=0#diff-b45bb451dbeb216c46cf560a6a8f26ef1ac5bdeb26d936a8fa0703aae326dde3R1-R5), [link](https://github.com/web-infra-dev/rspack/pull/3414/files?diff=unified&w=0#diff-fc80d9d773a38a2228ca4a3c3123924683d08ceef858a86c9fa12f4c4cde3183R1-R6))
*  Update test case outputs for electron main and preload targets ([link](https://github.com/web-infra-dev/rspack/pull/3414/files?diff=unified&w=0#diff-68435635c7abf06035e99ade8f46eeda6ad7614810533989afc5a1e324eb9db9R645-R649), [link](https://github.com/web-infra-dev/rspack/pull/3414/files?diff=unified&w=0#diff-68435635c7abf06035e99ade8f46eeda6ad7614810533989afc5a1e324eb9db9R719-R724))
*  Wrap exports in self-invoking function and pass require as parameter for JSONP chunk loading ([link](https://github.com/web-infra-dev/rspack/pull/3414/files?diff=unified&w=0#diff-3a6288d6314d317a4d96cddcf94d3fc655ce4bbade63cc83e05d5e4249025226L1-R1), [link](https://github.com/web-infra-dev/rspack/pull/3414/files?diff=unified&w=0#diff-3a6288d6314d317a4d96cddcf94d3fc655ce4bbade63cc83e05d5e4249025226L34-R37))
*  Add runtime functions and variables for resolving and loading chunks via JSONP ([link](https://github.com/web-infra-dev/rspack/pull/3414/files?diff=unified&w=0#diff-aa99eb877db60c83f83c419092e413c6bd69e4a0dc0298f83e5adb7458405c72R92-R131), [link](https://github.com/web-infra-dev/rspack/pull/3414/files?diff=unified&w=0#diff-aa99eb877db60c83f83c419092e413c6bd69e4a0dc0298f83e5adb7458405c72R149-R222), [link](https://github.com/web-infra-dev/rspack/pull/3414/files?diff=unified&w=0#diff-aa99eb877db60c83f83c419092e413c6bd69e4a0dc0298f83e5adb7458405c72R230-R234))
*  Remove module.exports assignment and wrap runtime code in second argument of self-invoking function ([link](https://github.com/web-infra-dev/rspack/pull/3414/files?diff=unified&w=0#diff-aa99eb877db60c83f83c419092e413c6bd69e4a0dc0298f83e5adb7458405c72L125-R459))

</details>

### CODE REVIEW

File: packages/rspack/src/config/zod/target.ts [hunk 1]
```diff
@@ -1,21 +1,59 @@
 import { z } from "zod";
 
-const allowTarget = z.enum([
-	"node",
```
Comment (hyf0) [inline packages/rspack/src/config/zod/target.ts:4]: I would suggest using string literal, which would infer a more accurate `type`. Users could use the type for auto-completion.

Comment (hyf0) [reply packages/rspack/src/config/zod/target.ts:4]: I'm ok with using `z.custom` on the `electron` part. I think it's better to list all the possible `es` targets, so people wouldn't write something like `es6`

Comment (ahabhgk) [reply packages/rspack/src/config/zod/target.ts:4]: you can write `es6` in webpack 🤷

Comment (ahabhgk) [reply packages/rspack/src/config/zod/target.ts:4]: Screenshot 2023-06-05 at 14 29 42
ok this is bad

**Review by hyf0 — DISMISSED** [review]

**Review by hardfist — APPROVED** [review]

**Pull request merged**